experimental/packages/sdk-logs/src/LogRecord.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Attributes, AttributeValue } from '@opentelemetry/api';
+import type * as logsAPI from '@opentelemetry/api-logs';
+import * as api from '@opentelemetry/api';
+import {
+  timeInputToHrTime,
+  isAttributeValue,
+  InstrumentationScope,
+} from '@opentelemetry/core';
+import type { IResource } from '@opentelemetry/resources';
+
+import type { ReadableLogRecord } from './export/ReadableLogRecord';
+import type { LogRecordLimits } from './types';
+import { Logger } from './Logger';
+
+export class LogRecord implements logsAPI.LogRecord, ReadableLogRecord {
+  time: api.HrTime;
+  severityText?: string;
+  severityNumber?: logsAPI.SeverityNumber;
+  body?: string;
+  readonly traceId?: string;
+  readonly spanId?: string;
+  readonly traceFlags?: number;
+  readonly resource: IResource;
+  readonly instrumentationScope: InstrumentationScope;
+  readonly attributes: Attributes = {};
+  private readonly _logRecordLimits: LogRecordLimits;
+
+  constructor(logger: Logger, logRecord: logsAPI.LogRecord) {
+    const {
+      timestamp = Date.now(),
+      severityNumber,
+      severityText,
+      body,
+      attributes = {},
+      spanId,
+      traceFlags,
+      traceId,
+    } = logRecord;
+
+    this.time = timeInputToHrTime(timestamp);
+    this.spanId = spanId;
+    this.traceId = traceId;
+    this.traceFlags = traceFlags;
+    this.severityNumber = severityNumber;
+    this.severityText = severityText;
+    this.body = body;
+    this.resource = logger.resource;
+    this.instrumentationScope = logger.instrumentationScope;
+    this._logRecordLimits = logger.getLogRecordLimits();
+    this.setAttributes(attributes);
+  }
+
+  public setAttribute(key: string, value?: AttributeValue) {
```

Comment:
LogRecord should be read-only after it has been emitted (see spec [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#onemit)).

> A LogRecordProcessor may freely modify logRecord for the duration of the OnEmit call. If logRecord is needed after OnEmit returns (i.e. for asynchronous processing) only reads are permitted.

I think this might need to be set in the logger after `onEmit` is called on the processor since there could be multiple processors.

Reply:
> LogRecord should be read-only after it has been emitted (see spec [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#onemit)).
> 
> > A LogRecordProcessor may freely modify logRecord for the duration of the OnEmit call. If logRecord is needed after OnEmit returns (i.e. for asynchronous processing) only reads are permitted.
> 
> I think this might need to be set in the logger after `onEmit` is called on the processor since there could be multiple processors.

fixed: make log record read-only after it has been emitted https://github.com/open-telemetry/opentelemetry-js/pull/3549/commits/f42369366ba5d2b7ceee4026a6901ba2391053cb